### DISCUSSION TITLE
[Bug] JVMService-produce high cpu 100%

### DISCUSSION THREAD

**Discussion by wangkang007** [apache/skywalking#9309] [Q&A]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

Java Agent (apache/skywalking-java)

### What happened

<img width="752" alt="image" src="https://user-images.githubusercontent.com/15212116/177289439-fea9f57b-275a-4175-946f-7b7f9f4c3118.png">

After the service runs for a period of time, SkywalkingAgent-5-JVMService-produce-0 will occupy 100% CPU

"SkywalkingAgent-5-JVMService-produce-0" #9 daemon prio=5 os_prio=0 cpu=199370183.56ms elapsed=606196.06s tid=0x00007feb31359800 nid=0x7c5 runnable [0x00007feae9e99000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000700399d28> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

### What you expected to happen

nothing

### How to reproduce

Not all services will happen, only one can be seen。

environment：
jdk: openjdk version "1.8.0_312"
runtime: k8s-container

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment]
We can't do an analysis from a CPU usage, especially you just post a stack showing queue blocked.

**Reply by wangkang007** [reply]
so What data do I need to provide here

**Reply by wangkang007** [reply]
https://github.com/apache/skywalking/discussions/8744 

There is a related discussion here, which is a little similar to this problem

**Reply by wu-sheng** [reply]
We gave a very clear response

**Diagnosis is not community's responsibility. You need to do research first.**

If you don't know what is happening, how should we?

**Reply by wu-sheng** [reply]
Community is not doing tech support. Don't trust us as a customer service provider. You have all the source codes, at least, you should use many existing tools to diagnose. Community could help if you have a certain question in a very clear scenario, but we wouldn't debug for you.

**Reply by wangkang007** [reply]
Well, I just feed back a problem. If you think you don't need to deal with it, it's OK, but I believe I'm not alone. I've seen some people meet it before.

**Comment by wangkang007** [comment]
@wu-sheng  hello wu，I think I found the problem，this code use cpu 100% ，when i have 5000 threads in my process and JDK < 17。 

<img width="906" alt="image" src="https://user-images.githubusercontent.com/15212116/178091106-d4b35b52-982c-429c-8c90-293b1e457658.png">



<img width="1520" alt="image" src="https://user-images.githubusercontent.com/15212116/178091144-c31d06c4-5475-4ff2-b425-03ec54cb6232.png">


<img width="713" alt="image" src="https://user-images.githubusercontent.com/15212116/178091185-34b27fcc-24e0-4783-bca4-5feeaca881a6.png">


As shown in the figure，When I have a lot of thread， getThreadMetrics will use more cpu。because jdk lookup thread info is slow。
if i hava 1000  thread，i will execute 1000*1000 count method to lookup thread info

**Reply by wu-sheng** [reply]
Again and again, why did you reas this thread?
1000 threads per JVM are totally wrong.

**Reply by wu-sheng** [reply]
I wouldn't reply to this thread anymore.

**Reply by wangkang007** [reply]
why 1000 threads per jvm are totally wrong？if i have 500 thread per jvm are totally wrong？I don't think this is a problem. We are all IO intensive, not CPU intensive

**Reply by wu-sheng** [reply]
Because it is pointless, you never have somany CPU kernel to support your application.

**Reply by wangkang007** [reply]
I am a web program，tomcat thread model is one request one thread。so i must have a lot of thread，Threads pass through time slices。so i can run program very well。

**Reply by wangkang007** [reply]
in tomcat thread model，when i hava 1 cpu，i can create 100 thread for web server

**Reply by wu-sheng** [reply]
If you say so. I am out.

**Reply by wangkang007** [reply]
bingo business is just waiting for io，because i will call many service and wait response，then i will process data。so i need so much thread

**Reply by wangkang007** [reply]
I don't want to adjust Tomcat or business threading model，This will bring great programming difficulty。It can't be said that I need to modify the threading model to adapt because of using SkyWalking. I believe someone will encounter such a situation in the future

**Reply by tivisGu** [reply]
> I don't want to adjust Tomcat or business threading model，This will bring great programming difficulty。It can't be said that I need to modify the threading model to adapt because of using SkyWalking. I believe someone will encounter such a situation in the future

Is there any temporary plan? In fact, I have also encountered this problem

**Reply by wangkang007** [reply]
> > I don't want to adjust Tomcat or business threading model，This will bring great programming difficulty。It can't be said that I need to modify the threading model to adapt because of using SkyWalking. I believe someone will encounter such a situation in the future
> 
> Is there any temporary plan? In fact, I have also encountered this problem

You can choose a temporary plan :

1.You can Upgrade jdk version to 11  to solve this problem perfectly。
2.If you use jdk 1.8, you need to try not to exceed 500 threads

**Reply by tivisGu** [reply]
Thanks a lot

**Comment by qingkong-x** [comment]
Maybe you can try: java -Dskywalking.jvm.metrics_collect_period=60
[https://skywalking.apache.org/docs/skywalking-java/v9.0.0/en/setup/service-agent/java-agent/configurations](https://skywalking.apache.org/docs/skywalking-java/v9.0.0/en/setup/service-agent/java-agent/configurations)

**Reply by wangkang007** [reply]
This may be a solution that can alleviate the problem, but it will result in a loss of accuracy
